File: 16833_HW2_EKF/code/EKF_slam/python/EKF_SLAM.py
import numpy as np
import re
import matplotlib.pyplot as plt
np.set_printoptions(suppress=True, threshold=np.inf, linewidth=np.inf)


def drawCovEllipse(c, cov, setting):
    """Draw the Covariance ellipse given the mean and covariance

    :c: Ellipse center
    :cov: Covariance matrix for the state
    :returns: None

    """
    U, s, Vh = np.linalg.svd(cov)
    a, b = s[0], s[1]
    vx, vy = U[0, 0], U[0, 1]
    theta = np.arctan2(vy, vx)
    R = np.array([[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]])
    phi = np.arange(0, 2*np.pi, np.pi/50)
    rot = []
    for i in range(100):
        rect = (np.array([3*np.sqrt(a)*np.cos(phi[i]), 3*np.sqrt(b)*np.sin(phi[i])]))[:, None]
        rot.append(R @ rect + c)

    rot = np.asarray(rot)
    plt.plot(rot[:, 0], rot[:, 1], c=setting, linewidth=0.75)


def drawTrajAndMap(X, last_X, P, t):
    """Draw Trajectory and map

    :X: Current state
    :last_X: Previous state
    :P: Covariance
    :t: timestep
    :returns: None

    """
    plt.ion()
    drawCovEllipse(X[0:2], P[0:2, 0:2], 'b')
    plt.plot([last_X[0], X[0]], [last_X[1], X[1]], c='b', linewidth=0.75)
    plt.plot(X[0], X[1], '*b')

    if t == 0:
        for k in range(6):
            drawCovEllipse(X[3 + k*2:3 + k*2+2], P[3 + k*2:3 + 2*k + 2, 3 + 2*k:3 + 2*k + 2], 'r')
    else:
        for k in range(6):
            drawCovEllipse(X[3 + k*2:3 + k*2+2], P[3 + 2*k:3 + 2*k + 2, 3 + 2*k:3 + 2*k + 2], 'g')

    plt.draw()
    plt.waitforbuttonpress(0)


def drawTrajPre(X, P):
    """ Draw trajectory for Predicted state and Covariance

    :X: Prediction vector
    :P: Prediction Covariance matrix
    :returns: None

    """
    drawCovEllipse(X[0:2], P[0:2, 0:2], 'm')
    plt.draw()
    plt.waitforbuttonpress(0)

def main():
    """Main function for EKF

    :arg1: TODO
    :returns: TODO

    """
    # TEST: Setup uncertainty parameters
    sig_x = 0.25
    sig_y = 0.1
    sig_alpha = 0.1
    sig_beta = 0.1
    sig_r = 0.16

    # Generate variance from standard deviation
    sig_x2 = sig_x**2
    sig_y2 = sig_y**2
    sig_alpha2 = sig_alpha**2
    sig_beta2 = sig_beta**2
    sig_r2 = sig_r**2

    # Open data file
    data_file = open("../../data/data.txt", 'r')

    # Read the first measurement data
    line = data_file.readline()
    fields = re.split('[\t ]', line)[:-1]
    arr = np.array([float(field) for field in fields])
    measure = arr[:, None]
    t = 1

    # Setup control and measurement covariance
    control_cov = np.diag([sig_x2, sig_y2, sig_alpha2])
    measure_cov = np.diag([sig_beta2, sig_r2])

    # Setup the initial pose vector and pose uncertainty
    pose = (np.array([0, 0, 0]))[:, None]
    pose_cov = np.diag([0.02**2, 0.02**2, 0.1**2])

    # TODO: Setup the initial landmark estimates landmark[] and covariance matrix landmark_cov[]
    # Hint: use initial pose with uncertainty and first measurement

    ##############################################################
    ################## Write your code here ######################
    ##############################################################

    # Setup state vector x with pose and landmark vector
    X = np.vstack((pose, landmark))

    # Setup covariance matrix P with pose and landmark covariance
    P = np.block([[pose_cov,           np.zeros((3, 2*k))],
                  [np.zeros((2*k, 3)),       landmark_cov]])

    # Plot initial state and covariance
    last_X = X
    drawTrajAndMap(X, last_X, P, 0)

    # Read file sequentially for controls
    # and measurements
    for line in data_file:
        fields = re.split('[\t ]', line)[:-1]
        arr = np.array([float(field) for field in fields])
        if arr.shape[0] == 2:
            d, alpha = arr[0], arr[1]
            control = (np.array([d, alpha]))[:, None]

            # TODO: Predict step
            # (Notice: predict state x_pre[] and covariance P_pre[] using input control data and control_cov[])

            ##############################################################
            ################## Write your code here ######################
            ##############################################################


            # Draw predicted state X_pre and Covariance P_pre
            drawTrajPre(X_pre, P_pre)

        # Read the measurement data
        else:
            measure = (arr)[:, None]

            # TODO: Correction step
            # (Notice: Update state X[] and covariance P[] using the input measurement data and measurement_cov[])

            ##############################################################
            ################## Write your code here ######################
            ##############################################################

            drawTrajAndMap(X, last_X, P, t)
            last_X = X
            t += 1

    # EVAL: Plot ground truth landmarks

    ##############################################################
    ################## Write your code here ######################
    ##############################################################



if __name__ == "__main__":
    main()

File: 16833_HW2_EKF/code/EKF_slam/matlab/EKF_SLAM.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  16833 Robot Localization and Mapping  %
%  Assignment #2                         %
%  EKF-SLAM                              %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear;
close all;
clc;

%==== TEST: Setup uncertainity parameters (try different values!) ===
% sig_x = 0.25;
% sig_y = 0.1;
% sig_alpha = 0.1;
% sig_beta = 0.01;
% sig_r = 0.08;
% % 
% sig_x = 0.5;
% sig_y = 0.3;
% sig_alpha = 0.3;
% sig_beta = 0.35;
% sig_r = 1.0;

sig_x = 2;
sig_y = 1;
sig_alpha = 1;
sig_beta = 0.35;
sig_r = 1.0;


%==== Generate sigma^2 from sigma ===
sig_x2 = sig_x^2;
sig_y2 = sig_y^2;
sig_alpha2 = sig_alpha^2;
sig_beta2 = sig_beta^2;
sig_r2 = sig_r^2;

%==== Open data file ====
fid = fopen('../../data/data.txt');

%==== Read first measurement data ====
tline = fgets(fid);
arr = str2num(tline);
measure = arr';
t = 1;

%==== Setup control and measurement covariances ===
control_cov = diag([sig_x2, sig_y2, sig_alpha2]);
measure_cov = diag([sig_beta2, sig_r2]);

%==== Setup initial pose vector and pose uncertainty ====
pose = [0 ; 0 ; 0];
pose_cov = diag([0.02^2, 0.02^2, 0.1^2]);

%==== TODO: Setup initial landmark vector landmark[] and covariance matrix landmark_cov[] ====
%==== (Hint: use initial pose with uncertainty and first measurement) ====

% Write your code here...
k=6;
l = [3,3,7,7,11,11; 
    6,12,8,14,6,12];
euclid =[];
maha = [];

landmark = zeros(2*k, 1);
landmark_cov=[];
for i=1:k
    beta_l = measure(2*(i-1)+1);
    r_l = measure(2*(i-1)+2);
    landmark(2*(i-1)+1 : 2*i , 1) = [pose(1) + r_l*cos(beta_l + pose(3)); 
                                    pose(2) + r_l*sin(beta_l + pose(3))];
    
    J_l = [1,0, -r_l*sin( pose(3) + beta_l ), cos(pose(3) + beta_l), -r_l*sin(pose(3) + beta_l);
            0,1, r_l*cos(pose(3) + beta_l), sin(pose(3) + beta_l), r_l*cos(pose(3) + beta_l)];
    
    cov_l = J_l * diag([diag(pose_cov)', diag(measure_cov)' ]) * J_l';
        
    landmark_cov = blkdiag(landmark_cov, cov_l); % TODO: do this better
end

%==== Setup state vector x with pose and landmark vector ====
x = [pose ; landmark];

%==== Setup covariance matrix P with pose and landmark covariances ====
P = [pose_cov zeros(3, 2*k) ; zeros(2*k, 3) landmark_cov];

%==== Plot initial state and conariance ====
last_x = x;
drawTrajAndMap(x, last_x, P, 0);

%==== Read control data ====
tline = fgets(fid);
while ischar(tline)
    arr = str2num(tline);
    d = arr(1);
    alpha = arr(2);

    %==== TODO: Predict Step ====
    %==== (Notice: predict state x_pre[] and covariance P_pre[] using input control data and control_cov[]) ====

    % Write your code here...
    [x_pre, P_pre] = calcPre(x, P, d, alpha, control_cov);

    %==== Draw predicted state x_pre[] and covariance P_pre[] ====
    drawTrajPre(x_pre, P_pre);
    pause(.01)

    %==== Read measurement data ====
    tline = fgets(fid);
    arr = str2num(tline);
    measure = arr';

    %==== TODO: Update Step ====
    %==== (Notice: update state x[] and covariance P[] using input measurement data and measure_cov[]) ====

    % Write your code here...
    for i_l = 1:2:length(arr)
        
        l_x = x(3 + i_l);
        l_y = x(3 + i_l + 1);
        d_rl = (l_x - x(1))^2 + (l_y - x(2))^2;
        
        H_p = [(l_y - x(2))/d_rl, -(l_x - x(1))/d_rl, -1;
                -(l_x - x(1))/sqrt(d_rl), -(l_y - x(2))/sqrt(d_rl), 0];
            
        H_l = [ -(l_y - x(2))/d_rl, (l_x - x(1))/d_rl;
                (l_x - x(1))/sqrt(d_rl), (l_y - x(2))/sqrt(d_rl)];

        K_pos = P_pre(1:3, 1:3) * H_p'/ (H_p * P_pre(1:3, 1:3) * H_p' + measure_cov);
        
        K_l = P_pre(3+i_l:3+i_l+1, 3+i_l:3+i_l+1) * H_l' / ...
                                ( H_l * P_pre(3+i_l:3+i_l+1, 3+i_l:3+i_l+1)*H_l' + measure_cov);
        
        meas_pred = [wrapToPi( atan2(l_y - x(2), l_x - x(1)) - x(3) ); sqrt(d_rl)];
        
        x(1:3) = x_pre(1:3) + K_pos * ( [arr(i_l); arr(i_l+1)] -  meas_pred );
        x(3+i_l:3+i_l+1) = x_pre(3+i_l:3+i_l+1) + K_l * ( [arr(i_l); arr(i_l+1)] -  meas_pred );
        
        P(1:3, 1:3) = (eye(3) - K_pos*H_p) * P_pre(1:3,1:3);
        P(3+i_l:3+i_l+1, 3+i_l:3+i_l+1) = (eye(2) - K_l*H_l) * P_pre(3+i_l:3+i_l+1, 3+i_l:3+i_l+1);
        
%         drawTrajAndMap(x, last_x, P, t);
    end
    %==== Plot ====    
    drawTrajAndMap(x, last_x, P, t);
    last_x = x;
    pause(.1)
    
    %== Distance stats ==
%     euclid_temp = zeros(1,k); maha_temp = zeros(1,k);
%     for i = 1:k
% %         beta_l = measure(2*(i-1)+1);
% %         r_l = measure(2*(i-1)+2);
%         
%         euclid_temp(1,i) = sqrt( ( x(3+2*(i-1)+1) - l(1,i) )^2 + ( x(3+2*(i-1)+2) - l(2,i) )^2 );
%         maha_temp(1,i) = sqrt( ([x(3+2*(i-1)+1) ;x(3+2*(i-1)+2) ] - l(:,i))' / ...
%                                 P(3+2*(i-1)+1:3+2*(i-1)+2, 3+2*(i-1)+1:3+2*(i-1)+2 ) ...
%                                 * ([x(3+2*(i-1)+1) ;x(3+2*(i-1)+2) ] - l(:,i)) );
%     end
    
%     euclid = [euclid; euclid_temp];
%     maha = [maha; maha_temp];
%     
    %==== Iteration & read next control data ===
    t = t + 1;
    tline = fgets(fid);    
end

for i = 1:k
    
    euclid(1,i) = sqrt( ( x(3+2*(i-1)+1) - l(1,i) )^2 + ( x(3+2*(i-1)+2) - l(2,i) )^2 );
    maha(1,i) = sqrt( ([x(3+2*(i-1)+1) ;x(3+2*(i-1)+2) ] - l(:,i))' / ...
                            P(3+2*(i-1)+1:3+2*(i-1)+2, 3+2*(i-1)+1:3+2*(i-1)+2 ) ...
                            * ([x(3+2*(i-1)+1) ;x(3+2*(i-1)+2) ] - l(:,i)) );
end

%==== EVAL: Plot ground truth landmarks ====

% Write your code here...

hold on
plot(l(1,:), l(2,:), 'k*')

%==== Close data file ====
fclose(fid);
